test/common/testlib.py

```diff
@@ -777,72 +834,191 @@ def copy_cores(self, title, label=None):
 def jsquote(str):
     return json.dumps(str)
 
-# See phantom-driver for the methods that are defined
-class Phantom:
-    def __init__(self, lang=None):
+class CDP:
+    def __init__(self, lang=None, headless=True):
         self.lang = lang
         self.timeout = 60
         self.valid = False
+        self.headless = headless
         self._driver = None
+        self._browser = None
+        self._browser_home = None
+        self._cdp_port_lockfile = None
 
     def __getattr__(self, name):
         if not name.startswith("_"):
-            return lambda *args: self._invoke(name, *args)
+            return lambda **kwargs: self._invoke(name, kwargs)
         raise AttributeError
 
-    def _invoke(self, name, *args):
+    def _invoke(self, name, kwargs):
+        trace = opts.trace and not kwargs.get("no_trace", False)
+        try:
+            del kwargs["no_trace"]
+        except KeyError:
+            pass
+
+        cmd = name.replace("_", ".") + "(" + json.dumps(kwargs) + ")"
```

Comment:
To be honest, I've never liked the way we used `__getattr__` in the `Phantom` class to call JavaScript functions. Replacing underscores by dots adds another level of magic and confusion for me. Something like

```python
cdp.invoke('Network.setCookie', **args)
```

reads much clearer to me. Do you feel strongly about this?

Reply:
No strong opinion about this, I just tried to stay close to what we have. But indeed `_invoke()` should be a private implementation detail of testlib.py, I don't think it leaks into the actual tests. So, I'm happy to change this.

Reply:
I'm fine if we do this in a follow-up as well (and I wouldn't mind doing it myself). As you're saying, it's quite unrelated.

Reply:
Fixed.